### ISSUE TITLE
Incorrect type deduction with sklearn

### ISSUE THREAD

**Issue by vimalk78** [microsoft/pyright#8974]
**Describe the bug**
Incorrect type deduction and hence incorrect flagging of type mismatch error.

**Code or Screenshots**

```python
from sklearn.datasets import fetch_california_housing

 california: Bunch = fetch_california_housing()
```

the error i see is
```
Diagnostics:
1. Type "tuple[Series | Unknown | DataFrame | ndarray[Any, dtype[float64]] | Any, Unknown | Series | DataFrame | NDArray[floating[Any]] | Any] | Bunch" is not assignable to declared type "Bunch"
     Type "tuple[Series | Unknown | DataFrame | ndarray[Any, dtype[float64]] | Any, Unknown | Series | DataFrame | NDArray[floating[Any]] | Any] | Bunch" is not assignable to type "Bunch"
       "tuple[Series | Unknown | DataFrame | ndarray[Any, dtype[float64]] | Any, Unknown | Series | DataFrame | NDArray[floating[Any]] | Any]" is not assignable to "Bunch" [reportAssignmentType]
```
The function `fetch_california_housing` returns a `Bunch`  type, but pyright thinks it returns some Tuple.

type as inferred by pyright:

```python
(function) def fetch_california_housing(
    *,
    data_home: Unknown | None = None,
    download_if_missing: bool = True,
    return_X_y: bool = False,
    as_frame: bool = False,
    n_retries: int = 3,
    delay: float = 1
) -> (tuple[Series | Unknown | DataFrame | ndarray[Any, dtype[float64]] | Any, Unknown | Series | DataFrame | NDArray[floating[Any]] | Any] | Bunch)
```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**NeoVim**

```
NVIM v0.10.1
Build type: Release
LuaJIT 2.1.1713484068
```

Using pyright in NeoVim via Mason.
pyright version: `1.1.380`

pyright config:

```
require 'lspconfig'.pyright.setup({
  on_attach = function(client, bufnr)
    -- Key mappings for LSP
    local nmap = function(keys, func, desc)
      if desc then
        desc = 'LSP: ' .. desc
      end
      vim.keymap.set('n', keys, func, { buffer = bufnr, desc = desc })
    end

    nmap('gd', require('telescope.builtin').lsp_definitions, '[G]oto [D]efinition')
    nmap('gr', require('telescope.builtin').lsp_references, '[G]oto [R]eferences')
    nmap('gI', require('telescope.builtin').lsp_implementations, '[G]oto [I]mplementation')
    nmap('<leader>D', require('telescope.builtin').lsp_type_definitions, 'Type [D]efinition')
    nmap('<leader>ds', require('telescope.builtin').lsp_document_symbols, '[D]ocument [S]ymbols')
    nmap('<leader>ws', require('telescope.builtin').lsp_dynamic_workspace_symbols, '[W]orkspace [S]ymbols')
  end,
  settings = {
    python = {
      analysis = {
        autoSearchPaths = true,
        diagnosticMode = "workspace",
        useLibraryCodeForTypes = true,
      },
    },
  },
})
```
scikit version

```
scikit-learn version: 1.5.2
```

**Comment by vimalk78** [comment]
let me add that there is no runtime error. these are all errors as seen in pyright.

**Comment by erictraut** [comment]
Could you please provide a self-contained, minimal code sample that demonstrates the problem you're seeing? The above code references symbols (e.g. `Bunch`) that are not defined.

The sklearn library is largely untyped — i.e. it has no type annotations. That means pyright falls back on type inference to attempt to determine type information. You may need to install (or create your own) type stubs to get the behavior you want. Alternatively, you can set `useLibraryCodeForTypes` to false in your pyright config. This tells pyright not to attempt to infer types from library code if the library is untyped.

**Comment by vimalk78** [comment]
`from sklearn.utils import Bunch`
`Bunch` is already imported in my code, i just didn't mention it in the issue.

perhaps pyright is already referring to type stubs? the type stub here is similar to the type detected by pyright, as it in it is a tuple
https://github.com/microsoft/python-type-stubs/blob/main/stubs/sklearn/datasets/_california_housing.pyi
```
def fetch_california_housing(
    *, data_home: None | str = None, download_if_missing: bool = True, return_X_y: bool = False, as_frame: bool = False
) -> tuple[Bunch, tuple]: ...
```

but the type returned by actual code is different
https://github.com/scikit-learn/scikit-learn/blob/1.5.2/sklearn/datasets/_california_housing.py#L241-L248
```
    return Bunch(
        data=X,
        target=y,
        frame=frame,
        target_names=target_names,
        feature_names=feature_names,
        DESCR=descr,
    )
```

**Comment by erictraut** [comment]
Are you using the type stubs that ship with pylance? Those stubs indicate that `fetch_california_housing` returns a value of type `tuple[Bunch, tuple]`. That type isn't compatible with the type `Bunch`, which is why you're seeing an error from pyright. If you think that the type information in those stubs are incorrect, please file a bug report with the pylance maintainers.

If you are not using the stubs that ship with pylance, then pyright will infer the return type from the library code based on the `return` statements in the function. In this case, the inferred return type is a union of multiple types including `Bunch` and a tuple. This inferred type is not assignable to `Bunch`, so pyright is reporting a type violation.

Either way, pyright is working correctly. It's telling you about a type violation in your code.

If you're not interested in seeing type-related errors, you can disable type checking diagnostics in pyright by setting `typeCheckingMode` to "off".

**Comment by vimalk78** [comment]
I used `useLibraryCodeForTypes` in `pyrightconfig.json`, but this seems to disable lsp, and i cannot browse the code anymore

**Comment by erictraut** [comment]
If you disable `useLibraryCodeForTypes`, pyright will no longer attempt to infer types for the code. Many language server features (like completion suggestions) require type information. If you primarily want language server features, you can leave `useLibraryCodeForTypes` set to true but disable type checking. If you want pyright to validate type correctness for your code, you'll need accurate type information for the library. You can get that by installing stubs or creating your own stubs. For more details, refer to [this documentation](https://microsoft.github.io/pyright/#/type-stubs).

**Comment by vimalk78** [comment]
I am relatively new to python, and data science libraries. Now i looked closely and see that `fetch_california_housing` does return different data types according to the input arguments. this was new to me.
Thank you for your time, and suggestion.
